### PR TITLE
Authorization support for multiple organizations

### PR DESCRIPTION
**Pull request by benwarfield-usds**
## Related Issue or Background Info

- in order to support multiple organizations (#405), we need the app to know what organization the current user belongs to 

## Changes Proposed

- add a claim to Okta to show a filtered list of groups that will map to the Organization external ID
- update the getCurrentOrganization method to examine this list and find the appropriate organization, if it exists (and probably turn the request away if not)

## Additional Information

- the proposed group-based model is described in [this comment on #405](/CDCgov/prime-simplereport/issues/405#issuecomment-751932928).
- this will also enable us to have a superuser role and an org-admin role, should we choose to start using them

### CODE REVIEW

**Review by timothybest-usds — APPROVED** [review]

**Comment by benwarfield-usds** [comment]
@pete-usds points out, ever so gently, that we should have new test cases in our integration tests to validate that users cannot (intentionally or otherwise) get data they shouldn't. I suspect that will need to wait for a separate PR, based on current timelines—need to do our best to validate manually instead.

File: ops/services/okta-global/main.tf [hunk 1]
```diff
@@ -48,15 +54,37 @@ resource "okta_auth_server_claim" "given_name" {
...
+  claim_type        = "RESOURCE"
+  name              = "devtest_roles"
+  value_type        = "GROUPS"
+  group_filter_type = "STARTS_WITH"
+  value             = "SR-TEST-"
+  scopes            = [
+    okta_auth_server_scope.sr_devtest.name]
 }
 
 // Create the CDC/USDS user groups
```
Comment (jeskew-gov) [inline ops/services/okta-global/main.tf:90]: The existing group ("Prime Team Members") doesn't correspond to any of the claims added in this PR. Is that group granted permissions somewhere outside of this file?

File: ops/services/okta-global/main.tf [hunk 2]
```diff
@@ -15,25 +15,31 @@ resource "okta_auth_server_scope" "sr" {
   default          = false
 }
 
-// Add the required claims
-resource "okta_auth_server_claim" "sr_user" {
```
Comment (jeskew-gov) [inline ops/services/okta-global/main.tf:19]: Were the `sr_user`, `sr_org`, and `groups` Okta claims being used by SR? I think this PR removes all of those.

File: ops/services/okta-global/main.tf [hunk 3]
```diff
@@ -65,6 +93,21 @@ resource "okta_group" "prime_users" {
...
+resource "okta_group" "simplereport_stg_admins" {
+  name        = "SR-STG-ADMINS"
+  description = "STAGING Application Administrators for SimpleReport"
+}
+
+resource "okta_group" "simplereport_devtest_admins" {
+  name        = "SR-TEST-ADMINS"
+  description = "DEV/TEST Application Administrators for SimpleReport"
+}
+
```
Comment (jeskew-gov) [inline ops/services/okta-global/main.tf:110]: Is the expectation that non-administrative groups (e.g., facility staff) gets managed within Okta or Simple Report somehow?

**Review by jeskew-gov — COMMENTED** [review]
Had some questions that mostly reflect my being new to the project and to Okta.

File: ops/services/okta-global/main.tf (hunk 1 above)
Comment (benwarfield-usds) [reply ops/services/okta-global/main.tf:90]: Yes, that group has login permission (an application assignment) granted manually in the console. It has no specific permissions beyond that. ￼

File: ops/services/okta-global/main.tf (hunk 2 above)
Comment (benwarfield-usds) [reply ops/services/okta-global/main.tf:19]: They are not (no claims are being used currently, just the app membership management).

File: ops/services/okta-global/main.tf (hunk 3 above)
Comment (benwarfield-usds) [reply ops/services/okta-global/main.tf:110]: Within Okta, manually for now and eventually via API calls from SR.

File: ops/services/okta-global/main.tf [hunk 4]
```diff
@@ -15,25 +15,31 @@ resource "okta_auth_server_scope" "sr" {
...
 
-// Add the required claims
-resource "okta_auth_server_claim" "sr_user" {
-  auth_server_id = data.okta_auth_server.default.id
-  claim_type     = "RESOURCE"
-  name           = "org"
-  value          = "appuser.pdi_org"
-  scopes = [
-  okta_auth_server_scope.sr.name]
+resource "okta_auth_server_scope" "sr_prod" {
```
Comment (nickrobison-usds) [inline ops/services/okta-global/main.tf:18]: Environment specific services should probably go in the persistent directory of each env.

File: ops/services/okta-global/main.tf [hunk 5]
```diff
@@ -48,15 +54,37 @@ resource "okta_auth_server_claim" "given_name" {
   value          = "user.firstName"
 }
 
-resource "okta_auth_server_claim" "groups" {
+resource "okta_auth_server_claim" "simplereport_prod_roles" {
```
Comment (nickrobison-usds) [inline ops/services/okta-global/main.tf:57]: This should be environment specific as well.

**Review by nickrobison-usds — COMMENTED** [review]
Overall, looks really good! I think there are some TF changes that would help improve things, but the logic looks sound to me.

File: ops/services/okta-app/main.tf [hunk 6]
```diff
@@ -78,3 +60,51 @@ resource "okta_app_group_assignment" "prime_users" {
     simple_report_user = null
   })
 }
+
+resource "okta_group" "simplereport_prod_admins" {
```
Comment (nickrobison-usds) [inline ops/services/okta-app/main.tf:64]: These should be collapsed into a single resource, like so:

```
resource "okta_group" "simplereport_admins" {
   name        = "SR-${upper(var.env)}-ADMINS"
   description = "${upper(var.env)} Application Administrators for SimpleReport"
 }
```

File: ops/services/okta-app/main.tf [hunk 7]
```diff
@@ -78,3 +60,51 @@ resource "okta_app_group_assignment" "prime_users" {
...
+  name        = "SR-STG-ADMINS"
+  description = "STAGING Application Administrators for SimpleReport"
+}
+
+resource "okta_group" "simplereport_dev_admins" {
+  name        = "SR-DEV-ADMINS"
+  description = "DEV/TEST Application Administrators for SimpleReport"
+}
+
+resource "okta_auth_server_claim" "simplereport_prod_roles" {
```
Comment (nickrobison-usds) [inline ops/services/okta-app/main.tf:79]: Same as the above pattern:

```
resource "okta_auth_server_claim" "simplereport_roles" {
   auth_server_id    = data.okta_auth_server.default.id
   claim_type        = "RESOURCE"
   name              = "${var.env}_roles"
   value_type        = "GROUPS"
   group_filter_type = "STARTS_WITH"
   value             = "SR-${upper(var.env)}-"
   scopes            = [
     okta_auth_server_scope.sr.name]
 }
```

File: ops/services/okta-global/main.tf [hunk 8]
```diff
@@ -15,25 +15,31 @@ resource "okta_auth_server_scope" "sr" {
...
 }
 
-resource "okta_auth_server_claim" "sr_org" {
-  auth_server_id = data.okta_auth_server.default.id
-  claim_type     = "RESOURCE"
-  name           = "access"
-  value          = "appuser.pdi_user"
-  scopes = [
-  okta_auth_server_scope.sr.name]
+resource "okta_auth_server_scope" "sr_stg" {
```
Comment (nickrobison-usds) [inline ops/services/okta-global/main.tf:26]: This should be moved into the `Okta-app` module as well. Using the variable pattern described earlier.

```
resource "okta_auth_server_scope" "sr" {
   auth_server_id   = data.okta_auth_server.default.id
   name             = "simple_report_${var.env}"
   description      = "${var.env}-only OAUTH scope for Simple Report application"
   metadata_publish = "NO_CLIENTS"
   default          = false
 }
```

**Review by nickrobison-usds — CHANGES_REQUESTED** [review]

**Pull request merged**